compiler/src/parsing/parsetree_iter.re

```diff
@@ -0,0 +1,486 @@
+  enter_foreign: (provide_flag, value_description) => unit,
+  leave_foreign: (provide_flag, value_description) => unit,
+  enter_primitive: (provide_flag, value_description) => unit,
+  leave_primitive: (provide_flag, value_description) => unit,
+  enter_top_let:
+    (provide_flag, rec_flag, mut_flag, list(value_binding)) => unit,
+  leave_top_let:
+    (provide_flag, rec_flag, mut_flag, list(value_binding)) => unit,
+  enter_module: (provide_flag, module_declaration) => unit,
+  leave_module: (provide_flag, module_declaration) => unit,
+  enter_pattern: pattern => unit,
+  leave_pattern: pattern => unit,
+  enter_expression: expression => unit,
+  leave_expression: expression => unit,
+  enter_type: parsed_type => unit,
+  leave_type: parsed_type => unit,
+  enter_toplevel_stmt: toplevel_stmt => unit,
+  leave_toplevel_stmt: toplevel_stmt => unit,
+  enter_constant: constant => unit,
+  leave_constant: constant => unit,
+  enter_let: (rec_flag, mut_flag, list(value_binding)) => unit,
+  leave_let: (rec_flag, mut_flag, list(value_binding)) => unit,
+  enter_value_binding: value_binding => unit,
+  leave_value_binding: value_binding => unit,
+  enter_data_declarations: list((provide_flag, data_declaration)) => unit,
+  leave_data_declarations: list((provide_flag, data_declaration)) => unit,
+  enter_data_declaration: data_declaration => unit,
+  leave_data_declaration: data_declaration => unit,
+};
+
+let iter_constant = (hooks, c) => {
+  hooks.enter_constant(c);
+  hooks.leave_constant(c);
+};
+
+let iter_location = (hooks, loc) => {
+  hooks.enter_location(loc);
+  hooks.leave_location(loc);
+};
+
+// Just a helper to extract the Location.t from a loc('a)
+let iter_loc = (hooks, {loc}) => {
+  iter_location(hooks, loc);
+};
+
+let iter_ident = (hooks, id) => {
+  open Identifier;
+  iter_loc(hooks, id);
+  let rec iter = id => {
+    switch (id) {
+    | IdentName(name) => iter_loc(hooks, name)
+    | IdentExternal(id, name) =>
+      iter(id);
+      iter_loc(hooks, name);
+    };
+  };
+  iter(id.txt);
+};
+
+let iter_attributes = (hooks, attrs) => {
+  List.iter(
+    ((attr_name, attr_args) as attr) => {
+      hooks.enter_attribute(attr);
+      iter_loc(hooks, attr_name);
+      List.iter(iter_loc(hooks), attr_args);
+      hooks.leave_attribute(attr);
+    },
+    attrs,
+  );
+};
+
+let rec iter_parsed_program = (hooks, {statements} as program) => {
+  hooks.enter_parsed_program(program);
+  iter_toplevel_stmts(hooks, statements);
+  hooks.leave_parsed_program(program);
+}
+
+and iter_toplevel_stmts = (hooks, stmts) => {
+  List.iter(iter_toplevel_stmt(hooks), stmts);
+}
+
+and iter_toplevel_stmt =
+    (hooks, {ptop_desc: desc, ptop_attributes: attrs, ptop_loc: loc} as top) => {
+  hooks.enter_toplevel_stmt(top);
+  iter_location(hooks, loc);
+  iter_attributes(hooks, attrs);
+  switch (desc) {
+  | PTopInclude(id) =>
+    hooks.enter_include(id);
+    iter_include(hooks, id);
+    hooks.leave_include(id);
```

Comment:
We've sort of got two different styles of who's calling `enter` and `leave`, sometimes it's the `iter` that's actually iterating the thing, and sometimes it's the thing that's calling that `iter`.

I think it'd be best to use just one style, and I'd be in favor of the iterator for the thing is the one that calls `enter` and `leave`. For example, `iter_toplevel_stmt` is the one responsible for calling `enter_toplevel_stmt` and `leave_toplevel_stmt`, so here, `iter_include` should internally call `enter_include` and `leave_include`, and the same for the rest of these.